### PR TITLE
Add simple asciinema.org recording

### PR DESCRIPTION
**Pull request by Jakob-Naucke**
This recording demonstrates the ease of deploying a confidential
container and its confidentiality through a memory check.

Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

Please also take this PR as a review of the recording itself.
This is just a link to the video with a thumbnail, which might be a bit controversial. In my (very personal) opinion, the video speaks for itself and a caption would just be fluff, but feel free to make counter-suggestions, also to the placement of the link :)
/cc @bpradipt @fidencio @fitzthum @jodh-intel @jiangliu @magowan @sameo @stevenhorsman

### CODE REVIEW

**Comment by Jakob-Naucke** [comment]
and /cc @c3d, thanks for the idea :)

**Comment by c3d** [comment]
Thanks @Jakob-Naucke for the nice demo. Very interesting, short and convincing. Certainly convincing enough for the community, maybe a bit fast for people just joining.

A couple of quick observations (with times), to consider should you decide to re-record other demos some day. *Please do not re-record for these remarks, the video is good as is*

In general, it might be good to draw attention to some highlights (see a [couple of ways of doing that ](https://asciinema.org/a/xgZYqLMmueyXdNM0fHsy3ZHwt)with `grep or in `tmux` with`C-b [` and `C-space` - You probably want to have mouse enabled in `tmux` for that to be usable.

* `00:00` You start with a `tmux a` which is a bit distracting. What about `asciinema rec foo.cast -c tmux` ?
* `00:02` The contents of the two files are important. It might be worth showing the `less` command that goes with each so that we understand what they are. Or add a comment in each file, something like `# A regular Alpine container` and `# A confidential container with a service providing ssh support`. Also, as is, the `ccv0` looks much more complex even if, in reality, it's the addition of a service and deployment that generate the added complexity. Maybe making the "regular" side symmetric by also having a service and deployment (see why below).
* `00:08` It looks to me like you are comparing a `kc exec` with an `ssh`, since `kc exec` is blocked on the CC side. Problem is that you could imagine it's actually `kc exec` being insecure and `ssh` protecting the data. So that's why also having an `ssh` on the non-CC side would show that this is not the case.
* `00:13` You are scrolling the CC yaml file while the non-CC pod is starting up. I tried to figure out what you were trying to say, and I believe that you are saying nothing there 😉 since in the end you return to the non-CC case. If you are trying to highlight something, I did not understand what it was. I think it would be better to have that side just show a shell prompt until you actually are ready to show something.
* `00:22` Would you like using `qmp-shell` for clarity instead of your `socat` magic? Or wrap it in your own `dump-guest-memory` script (which could show the commands, the point being that if I see you type `dump-guest-memory`, I have an idea what you are trying to do without having to read all the way to the right)? BTW, had to scratch my head to guess why you were using `compgen -G` here, so it's a bit distracting 🤔 
* `00:37` You could split the `kc exec` on its one command line, so that we understand that it's intentional for this command to be blocked. As is, it looks like your `ccv0` is just ending with an `Internal error`. We know it's good, newbies might not 😄  Maybe highlight the error to show it's intentional.
* `00:50`: Highlighting the secrets
* `00:51`: It looks like you have the same prompt in your host and in your containers, so with the `$` prompt it was hard for me to tell if the second `socat` was done from within the container.
* `00:54`: The end game is perfect. Top says `Retrieved confidential data`, bottom says `Found no confidential data`, that's really cool. If you like emoji, remember that modern terminals actually support the real ones, so it could look `Retrieved confidential data 🙄` and `Found no confidential data ✅` 😄 

Anyway, great job as is, and I care much more about what is happening under the hood…

**Review by c3d — APPROVED** [review]
Really cool, thanks!

**Comment by Jakob-Naucke** [comment]
@c3d thank you for your feedback!

> A couple of quick observations (with times), to consider should you decide to re-record other demos some day. 

I do want to record a longer demo that will also be more specific to Secure Execution. Most things you address won't directly apply to that (the format will be "~5 minutes with voiceover" instead of "as fast as possible with only text communication"), but it's valuable anyway.

> * `00:00` You start with a `tmux a` which is a bit distracting. What about `asciinema rec foo.cast -c tmux` ?

yeah, first time using asciinema :) I should have checked `asciinema rec --help` instead of `asciinema --help`.

> * `00:02` The contents of the two files are important. It might be worth showing the `less` command that goes with each so that we understand what they are. Or add a comment in each file, something like `# A regular Alpine container` and `# A confidential container with a service providing ssh support`.

My intention was to have the filenames suffice. However, for the standard container, I just noticed `less` doesn't even display that. Yeah, could be better.

> Also, as is, the `ccv0` looks much more complex even if, in reality, it's the addition of a service and deployment that generate the added complexity. Maybe making the "regular" side symmetric by also having a service and deployment (see why below).
> * `00:08` It looks to me like you are comparing a `kc exec` with an `ssh`, since `kc exec` is blocked on the CC side. Problem is that you could imagine it's actually `kc exec` being insecure and `ssh` protecting the data. So that's why also having an `ssh` on the non-CC side would show that this is not the case.

I had considered using SSH for the non-CC as well, but decided against it for the sake of showing the "normal", more familiar way to get a container shell. (Well, in production k8s, you probably won't be running shells anyway. Still.) I see your point, but I think it's still pretty clear to anyone familiar with "protection of data at rest/in motion/in use", for they can see that the dump is one of memory, not of socket/net traffic.

> * `00:13` You are scrolling the CC yaml file while the non-CC pod is starting up. I tried to figure out what you were trying to say, and I believe that you are saying nothing there :wink: since in the end you return to the non-CC case. If you are trying to highlight something, I did not understand what it was. I think it would be better to have that side just show a shell prompt until you actually are ready to show something.

This was a crude attempt at shortening the video through bridging some time there which didn't land very well, I admit.

> * `00:22` Would you like using `qmp-shell` for clarity instead of your `socat` magic? 

Using `socat` instead of `qmp-shell` was mostly setup laziness, yeah :)

> Or wrap it in your own `dump-guest-memory` script (which could show the commands, the point being that if I see you type `dump-guest-memory`, I have an idea what you are trying to do without having to read all the way to the right)?

That's an idea. It should definitely show commands though, I want to avoid any "oh your script does some hidden magic" accusations. (Well, in principle, you can't really prove this isn't faked anyway. Still.)

> BTW, had to scratch my head to guess why you were using `compgen -G` here, so it's a bit distracting :thinking:

This expands `/run/vc/vm/${some_long_pod_hash}/…`, which works as long as you have only one pod. The only alternative I'm aware of is some `crictl pods -q --no-trunc --name ccv0-ssh.*` magic (well, that actually turned out to be shorter than I had thought. Anyway.)

> * `00:51`: It looks like you have the same prompt in your host and in your containers, so with the `$` prompt it was hard for me to tell if the second `socat` was done from within the container.

It isn't. The `ssh` that I run is not interactive, it only sets the variable. The container prompt would indeed look different.

> Certainly convincing enough for the community, maybe a bit fast for people just joining. [...] _Please do not re-record for these remarks, the video is good as is_ 

Thank you for reassuring me. Yes, I think with operator integration for non-TEE, Secure Execution, a video for that, and further documentation, I don't think a re-recording is a right-now priority :)

**Comment by Jakob-Naucke** [comment]
For full disclosure's sake, I used this diff in Kata:
```diff
diff --git a/src/runtime/virtcontainers/qemu.go b/src/runtime/virtcontainers/qemu.go
index 29cd0d17..1ebe0cad 100644
--- a/src/runtime/virtcontainers/qemu.go
+++ b/src/runtime/virtcontainers/qemu.go
@@ -382,6 +382,12 @@ func (q *qemu) createQmpSocket() ([]govmmQemu.QMPSocket, error) {
                        Server: true,
                        NoWait: true,
                },
+               {
+                       Type:   "unix",
+                       Name:   q.qmpMonitorCh.path + ".extra",
+                       Server: true,
+                       NoWait: true,
+               },
        }, nil
 }
 
```

File: README.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 # Welcome to documentation repository for Confidential Containers 
 
+[![asciicast](https://asciinema.org/a/eGHhZdQY3uYnDalFAfuB7VYqF.svg)](https://asciinema.org/a/eGHhZdQY3uYnDalFAfuB7VYqF)
```
Comment (magowan) [inline README.md:5]: Not sure if it is my setup, but this appears quite large and results in the following text not being on my screen?
I wonder if the embedded demo would be better  little lower down , just above the Further Detail heading?
That way the reader has a chance to get context then see the demo as they scroll down?

Comment (Jakob-Naucke) [reply README.md:5]: > Not sure if it is my setup, but this appears quite large and results in the following text not being on my screen?

The rich diff view might look weird, does [this](https://github.com/Jakob-Naucke/documentation/tree/asciinema-ssh) look better?

> I wonder if the embedded demo would be better little lower down , just above the Further Detail heading? That way the reader has a chance to get context then see the demo as they scroll down?

Maybe better, let me move it.

Comment (magowan) [reply README.md]: I think having moved it to Further Detail section works for me.

**Review by magowan — APPROVED** [review]

**Comment by bpradipt** [comment]
@Jakob-Naucke  really cool demo. 
A question on the location of the demo video - should this be added to the ssh-demo readme or the main readme page or both?

**Comment by Jakob-Naucke** [comment]
> A question on the location of the demo video - should this be added to the ssh-demo readme or the main readme page or both?

IMO, it belongs in the main readme, and not in the ssh-demo readme. The video isn't really about the SSH demo, it's about the ease of creating a confidential container, and SSH is merely a vehicle there (real workloads are probably non-interactive).

**Review by fitzthum — APPROVED** [review]
LGTM. We could go back and forth about exactly the most concise and intuitive way to show the project (and we probably should), but this seems like a good starting point.

**Review by sameo — APPROVED** [review]
Really nice, thanks a lot @Jakob-Naucke

**Pull request merged**